src/vs/workbench/contrib/notebook/browser/contrib/debug/notebookDebugDecorations.ts

```diff
@@ -0,0 +1,143 @@
+				this.onDidChangeCallStack();
+			}
+		}));
+	}
+
+	private onDidChangeCallStack(): void {
+		const exes = this._notebookExecutionStateService.getCellExecutionsByHandleForNotebook(this._notebookEditor.textModel!.uri);
+
+		const topFrameCellsAndRanges: ICellAndRange[] = [];
+		let focusedFrameCellAndRange: ICellAndRange | undefined = undefined;
+
+		const getNotebookCellAndRange = (sf: IStackFrame): ICellAndRange | undefined => {
+			const parsed = CellUri.parse(sf.source.uri);
+			if (parsed && parsed.notebook.toString() === this._notebookEditor.textModel?.uri.toString()) {
+				return { handle: parsed.handle, range: sf.range };
+			}
+			return undefined;
+		};
+
+		for (const session of this._debugService.getModel().getSessions()) {
+			for (const thread of session.getAllThreads()) {
+				const topFrame = thread.getTopStackFrame();
+				if (topFrame) {
+					const notebookCellAndRange = getNotebookCellAndRange(topFrame);
+					if (notebookCellAndRange) {
+						topFrameCellsAndRanges.push(notebookCellAndRange);
+						exes?.delete(notebookCellAndRange.handle);
+					}
+				}
+			}
+		}
+
+		const focusedFrame = this._debugService.getViewModel().focusedStackFrame;
+		if (focusedFrame && focusedFrame.thread.stopped) {
+			const thisFocusedFrameCellAndRange = getNotebookCellAndRange(focusedFrame);
+			if (thisFocusedFrameCellAndRange &&
+				!topFrameCellsAndRanges.some(topFrame => topFrame.handle === thisFocusedFrameCellAndRange?.handle && Range.equalsRange(topFrame.range, thisFocusedFrameCellAndRange?.range))) {
+				focusedFrameCellAndRange = thisFocusedFrameCellAndRange;
+				exes?.delete(focusedFrameCellAndRange.handle);
+			}
+		}
+
+		this.setTopFrameDecoration(topFrameCellsAndRanges);
+		this.setFocusedFrameDecoration(focusedFrameCellAndRange);
+
+		const exeHandles = exes ?
+			Array.from(exes.entries())
+				.filter(([_, exe]) => exe.state === NotebookCellExecutionState.Executing)
+				.map(([handle]) => handle)
+			: [];
+		this.setExecutingCellDecorations(exeHandles);
+	}
+
+	private setTopFrameDecoration(handlesAndRanges: ICellAndRange[]): void {
+		const newDecorations = handlesAndRanges.map(({ handle, range }) => {
+			const options: INotebookCellDecorationOptions = {
+				overviewRuler: {
+					color: topStackFrameColor,
+					includeOutput: false,
+					modelRanges: [range],
+					position: NotebookOverviewRulerLane.Full
+				}
+			};
+			return { handle, options };
+		});
+
+		this._currentTopDecorations = this._notebookEditor.deltaCellDecorations(this._currentTopDecorations, newDecorations);
+	}
+
+	private setFocusedFrameDecoration(focusedFrameCellAndRange: ICellAndRange | undefined): void {
+		let newDecorations: INotebookDeltaDecoration[] = [];
+		if (focusedFrameCellAndRange) {
+			const options: INotebookCellDecorationOptions = {
+				overviewRuler: {
+					color: focusedStackFrameColor,
+					includeOutput: false,
+					modelRanges: [focusedFrameCellAndRange.range],
+					position: NotebookOverviewRulerLane.Full
```

Comment:
Usually decorations are rendered in single lane (left, center or right), do they (debugging) have to be "Full"?

Reply:
It is for the normal editor debugging decoration, I'm copying from there.